java/code/src/com/redhat/rhn/manager/system/test/SystemManagerTest.java

```diff
@@ -1982,6 +1982,87 @@ public void testCreateProxyContainerConfigExisting() throws InstantiationExcepti
         testCreateProxyContainerConfig();
     }
 
+
+    /**
+     * Tests creating multiple chained proxy container configuration file.
+     * Network topology: PXY0 -> PXY1 -> PXY2A, PXY2B
+     * PXY0 is an anchor proxy;
+     * PXY1 parent/server is PXY0;
+     * Both PXY2A and PXY2B parent/server is PXY1;
+     *
+     * @throws InstantiationException
+     * @throws IOException
+     */
+    @Test
+    public void testCreateChainedProxyContainerConfig() throws InstantiationException, IOException {
+        user.addPermanentRole(RoleFactory.ORG_ADMIN);
+        final long maxCache = 4096;
+        final String email = "admin@mgr.lab";
+        final String rootCA = "Dummy Root CA";
+        final List<String> otherCAs = List.of("CA 1", "CA 2");
+        final String cert = "Dummy cert";
+        final String key = "Dummy key";
+        final String apacheCert = "Dummy cert for apache";
+
+        final String pxy0Name = "pxy0.mgr.lab";
+        final String pxy1Name = "pxy1.mgr.lab";
+        final String pxy2AName = "pxy2a.mgr.lab";
+        final String pxy2BName = "pxy2b.mgr.lab";
+        final String pxy0PubKey = "SshPublicKey " + pxy0Name;
+        final String pxy1SshPushKey = "sshPushKey" + pxy1Name;
+        final String pxy1SshPushPubKey = "sshPushPubKey" + pxy1Name;
+        final String pxy2ASshPushKey = "sshPushKey" + pxy2AName;
+        final String pxy2ASshPushPubKey = "sshPushPubKey" + pxy2AName;
+        final String pxy2BSshPushKey = "sshPushKey" + pxy2BName;
+        final String pxy2BSshPushPubKey = "sshPushPubKey" + pxy2BName;
+
+        createTestProxy(pxy0Name);
+
+        context().checking(new Expectations() {{
+            allowing(saltServiceMock).generateSSHKey(with(aNull(String.class)), with(aNull(String.class)));
+            will(onConsecutiveCalls(
+                    returnValue(Optional.of(new MgrUtilRunner.SshKeygenResult(pxy1SshPushKey, pxy1SshPushPubKey))),
+                    returnValue(Optional.of(new MgrUtilRunner.SshKeygenResult(pxy2ASshPushKey, pxy2ASshPushPubKey))),
+                    returnValue(Optional.of(new MgrUtilRunner.SshKeygenResult(pxy2BSshPushKey, pxy2BSshPushPubKey)))
+            ));
+            allowing(saltServiceMock)
+                    .checkSSLCert(with(equal(rootCA)), with(equal(new SSLCertPair(cert, key))), with(equal(otherCAs)));
+            will(returnValue(apacheCert));
```

Comment:
Please rebase on master: this unit test will need some more expectations as I enhanced the `SSLCertManager` recently